exonum/src/blockchain/block.rs

```diff
@@ -349,18 +356,117 @@ impl IndexProof {
 
         // The index proof should feature exactly one present entry.
         let mut unchecked_entries = self.index_proof.all_entries_unchecked();
-        let (name, maybe_hash) = unchecked_entries.next().ok_or(ProofError::NoIndex)?;
+        let (name, maybe_hash) = unchecked_entries.next().ok_or(ProofError::NoEntry)?;
         if unchecked_entries.next().is_some() {
-            return Err(ProofError::AmbiguousIndex);
+            return Err(ProofError::AmbiguousEntry);
         }
-        let index_hash = *maybe_hash.ok_or(ProofError::NoIndex)?;
+        let index_hash = *maybe_hash.ok_or(ProofError::NoEntry)?;
         self.index_proof
             .check_against_hash(self.block_proof.block.state_hash)
-            .map_err(ProofError::IncorrectIndexProof)?;
+            .map_err(ProofError::IncorrectEntryProof)?;
         Ok((name.as_str(), index_hash))
     }
 }
 
+/// Proof of authenticity for a single top-level call in a block, such as a [transaction].
+///
+/// The proof consists of two parts:
+///
+/// - `block_proof`: block header with the proof of authenticity
+/// - `call_proof`: proof from the error aggregator (i.e., a `ProofMapIndex` the Merkle root
+///   of which is recorded in the block header as `error_hash`).
+///
+/// For an execution that resulted in an error, `call_proof` will be a proof of existence.
+/// If a transaction was executed successfully, `call_proof` will be a proof of absence.
+/// Since the number of transactions in a block is mentioned in the block header, the user
+/// will be able to distinguish absence of error (meaning successful execution) from
+/// the absence of a transaction with such an index. Indeed, if the index is less
+/// than amount of transactions in block, the proof denotes successful execution;
+/// otherwise, the transaction with the given index does not exist in the block.
+///
+/// Similarly, execution errors of the `before_transactions` / `after_transactions`
+/// hooks can be proven to external clients. Discerning successful execution
+/// from a non-existing service requires prior knowledge though.
+///
+/// `CallProof`s should not be mixed up with a proof of transaction commitment.
+/// To verify that a certain transaction was committed, use a proof from
+/// the `block_transactions` index of the [core schema].
+///
+/// # Verification
+///
+/// If you are implementing `CallProof` verification, bear in mind that `ExecutionError`
+/// need to be stripped off the non-hashed information before verifying `call_proof`.
```

Comment:
> If the proof, on the other hand, comes only with the data to be hashed

Are we OK with not returning human-readable error description? IMO, the complexity introduced by error descriptions is manageable, but another option is to return error descriptions separately. Would it be OK with you?

Reply:
I think the human-readable error description is valuable. Do you mean by returning separately returning in the same response, but in a different object? I would go with such approach, because it does not require special handling during verification.